tardis/utilities/asyncbulkcall.py

```diff
@@ -0,0 +1,173 @@
+from typing import TypeVar, Generic, Iterable, List, Tuple, Optional, Union
+from typing_extensions import Protocol
+import asyncio
+import time
+import sys
+
+from backports.cached_property import cached_property
+
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+class BulkCommand(Protocol[T, R]):
+    """
+    Protocol of callables suitable for :py:class:`~.BulkExecution`
+
+    A bulk command must take an arbitrary number of tasks and is expected to provide
+    an iterable of one result per task. Alternatively, it may provide a single
+    :py:data:`None` value to indicate that there is no result. An unhandled
+    :py:class:`Exception` means that all tasks failed with that :py:class:`Exception`.
+    """
+
+    async def __call__(self, *__tasks: T) -> Optional[Iterable[R]]:
+        ...
+
+
+class AsyncBulkCall(Generic[T, R]):
+    """
+    Framework for queueing and executing several tasks via bulk commands
+
+    :param command: async callable that executes several tasks
+    :param size: maximum number of tasks to execute in one bulk
+    :param delay: maximum age in seconds of tasks before executing them
```

Comment:
Given the two parameters `size` and `delay` it currently does not become clear that both are used to trigger the creation of bulks. Especially the parameter `size` does have two meanings: i) the maximum size of a bulk and ii) the maximum length of the queue until new bulks are created.
Please consider making it clear in the docstring.

```suggestion
    :param size: maximum number of tasks before executing them in bulks of this maximum size
    :param delay: maximum age in seconds of tasks before executing them
```